recipes/cpython/all/conanfile.py

```diff
@@ -380,18 +381,29 @@ def package(self):
             if self._is_py2:
                 scripts = ("2to3", "idle", "pydoc", "python{}-config".format(self._version_major_minor), "smtpd.py")
             else:
-                script_prefixes = ("2to3-", "easy_install-", "idle", "pydoc", "pyvenv-", "pip")
+                script_prefixes = ("2to3-", "easy_install-", "idle", "pydoc", "pyvenv-", "pip*")
                 scripts = tuple(os.path.join(self.package_folder, "bin", sp + self._version_major_minor) for sp in script_prefixes)
 
             for script_name in scripts:
                 script = os.path.join(self.package_folder, "bin", script_name)
                 if not os.path.isfile(script):
                     continue
+                if os.path.islink(script):
+                    continue
                 with open(script, "r") as fn:
                     fn.readline()
                     text = fn.read()
                 with open(script, "w") as fn:
-                    fn.write("""#!/bin/sh\n''':'\n__file__="$0"; while [ -L "$__file__" ]; do __file__="$(dirname "$__file__")/$(readlink "$__file__")"; done; exec "$(dirname "$__file__")/python{}" "$0" "$@"\n'''\n""".format(self._version_major_minor))
+                    fn.write(textwrap.dedent("""\
+                        #!/bin/sh
+                        ''':'
+                        __file__="$0";
+                        while [ -L "$__file__" ]; do
+                            __file__="$(dirname "$__file__")/$(readlink "$__file__")";
+                        done;
+                        exec "$(dirname "$__file__")/python{}" "$0" "$@"
```

Comment:
@ohanar 
Are you sure the `$@` must be surrounded with double quotes?

Reply:
Yes you do. I agree it is weird. Basically it is to prevent the shell from re-parsing the arguments. Consider the following:
```sh
$ echo -e '#!/bin/sh\nexec python "$@"' > /tmp/python.sh
$ sh /tmp/python.sh -c 'print("Hello World!")'
Hello World!
$ echo -e '#!/bin/sh\nexec python $@' > /tmp/python.sh # No double quotes now
$ sh /tmp/python.sh -c 'print("Hello World!")'
  File "<string>", line 1
    print("Hello
               ^
SyntaxError: EOL while scanning string literal
```
What's happening when you don't quote the `$@` is that the space in the hello world is getting re-parsed, and so python sees three arguments: `-c`, `print("Hello`, and `World!")`. We want to perfectly forward the arguments, so we use the double quotes, and python gets the two arguments that we would expect: `-c` and `print("Hello World!")`.

Reply:
Ah okay it replaces itself with a subshell that will execute the commands.
I'm now reading the documentation of exec. It reads:
```
       exec [-cl] [-a name] [command [arguments]]
              If command is specified, it replaces the shell.  No new process is created.  The arguments become the
              arguments to command.  If the -l option is supplied, the shell places a dash at the beginning of  the
              zeroth  argument  passed to command.  This is what login(1) does.  The -c option causes command to be
              executed with an empty environment.  If -a is supplied, the shell passes name as the zeroth  argument
              to  the  executed  command.   If  command cannot be executed for some reason, a non-interactive shell
              exits, unless the execfail shell option is enabled.  In that case, it returns failure.   An  interac‐
              tive  shell returns failure if the file cannot be executed.  A subshell exits unconditionally if exec
              fails.  If command is not specified, any redirections take effect  in  the  current  shell,  and  the
              return status is 0.  If there is a redirection error, the return status is 1.
```
Shouldn't this line then become:
```sh
exec -a "$0" "$(dirname "$__file__")/python{}" "$@"
``
?

Reply:
Also, you don't need trailing semicolons. Newlines serve the same purpose as semicolons.

Reply:
Ah yes, good catch. I was too focused on the reformatting :smile:

Reply:
> Shouldn't this line then become:
> 
> ```shell
> exec -a "$0" "$(dirname "$__file__")/python{}" "$@"
> ```
> 
> ?

That looks like some bash extension. See [here](https://pubs.opengroup.org/onlinepubs/009695399/utilities/exec.html) for the posix documentation.

Reply:
Indeed, copied from the bash manual.

Should this line then become the following? (no `$0`)
```sh
exec "$(dirname "$__file__")/python{}" "$@"
```

Reply:
You definitely want the $0, $@ does not contain the zeroth argument (i.e. the script as named by the user). Without the $0, the original script will not be launched, so all the scripts would more or less be equivalent to symlinks to the python executable.

Reply:
Basically, that line is converting something like `/usr/bin/pip3 install wheel` to something like `/usr/bin/python /usr/bin/pip3 install wheel`, which is functionally equivalent.

Reply:
Okay, sorry. I forget the main picture for a moment.
We're all good now :smile: